tools/graph-stats/graph-memory-stats.cpp

```diff
@@ -0,0 +1,308 @@
+#include <iostream>
+#include <string>
+#include <unordered_map>
+
+#include <arrow/type_traits.h>
+
+#include "katana/ArrowVisitor.h"
+#include "katana/Galois.h"
+#include "katana/JSON.h"
+#include "katana/Logging.h"
+#include "katana/OfflineGraph.h"
+#include "llvm/Support/CommandLine.h"
+#include "tsuba/Errors.h"
+#include "tsuba/FaultTest.h"
+#include "tsuba/FileView.h"
+
+namespace cll = llvm::cl;
+static cll::opt<std::string> inputfilename(
+    cll::Positional, cll::desc("graph-file"), cll::Required);
+
+static cll::opt<std::string> outputfilename(
+    cll::Positional, cll::desc("out-file"), cll::Required);
+
+using map_element = std::unordered_map<std::string, int64_t>;
+using map_string_element = std::unordered_map<std::string, std::string>;
+using memory_map = std::unordered_map<
+    std::string, std::variant<map_element, map_string_element>>;
+
+struct Visitor : public katana::ArrowVisitor {
+  using ResultType = katana::Result<std::pair<int64_t, int64_t>>;
+  using AcceptTypes = std::tuple<katana::AcceptAllFlatTypes>;
+
+  template <typename ArrowType, typename ArrayType>
+  arrow::enable_if_null<ArrowType, ResultType> Call(const ArrayType& scalars) {
+    fmt::print("{} : {}\n", scalars.type()->ToString(), scalars.length());
+    return std::pair(0, 0);
+  }
+
+  template <typename ArrowType, typename ArrayType>
+  std::enable_if_t<
+      arrow::is_number_type<ArrowType>::value ||
+          arrow::is_boolean_type<ArrowType>::value ||
+          arrow::is_temporal_type<ArrowType>::value,
+      ResultType>
+  Call(const ArrayType& scalars) {
+    using mainType = typename ArrayType::TypeClass;
+    int64_t real_used_space = 0;
+    for (auto j = 0; j < scalars.length(); j++) {
+      if (!scalars.IsNull(j)) {
+        real_used_space += sizeof(mainType);
+      }
+    }
+
+    int64_t space_allocated = sizeof(mainType) * scalars.length();
+    return std::pair(space_allocated, real_used_space);
+  }
+
+  template <typename ArrowType, typename ArrayType>
+  arrow::enable_if_string_like<ArrowType, ResultType> Call(
+      const ArrayType& scalars) {
+    using widthType = typename ArrayType::offset_type;
+    int64_t total_width = scalars.total_values_length();
+    int64_t metadata_size = sizeof(widthType) * scalars.length();
+    return std::pair(total_width + metadata_size, total_width);
+  }
+
+  ResultType AcceptFailed(const arrow::Array& scalars) {
+    return KATANA_ERROR(
+        katana::ErrorCode::ArrowError, "no matching type {}",
+        scalars.type()->ToString());
+  }
+};
+
+void
+PrintAtomicTypes(const std::vector<std::string>& atomic_types) {
+  for (auto atype : atomic_types) {
+  }
+}
+
+template <typename T>
+void
+PrintMapping(const std::shared_ptr<T> u) {
+  for (const auto& n : u) {
+    fmt::print("{} : {}\n", n->first, n->second);
+  }
+}
+
+std::pair<int64_t, int64_t>
+RunVisit(
+    const std::shared_ptr<arrow::Array> scalars, const std::string name,
+    const std::shared_ptr<map_element> allocations,
```

Comment:
why do you need a `shared_ptr<map_element>` everywhere? `shared_ptr` is useful when you don't know who or when exactly you'll free dynamically allocated memory. If ownership and lifetime are clear then use `unique_ptr`. In this case, I suspect you need neither. A simple pointer will suffice.